File: ilRandom/ilRand.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Diploma2
{
    [Serializable]
    public class ilRand
    {
        private const ulong m = 1442695040888963407;
        private const ulong a = 6364136223846793005;

        private ulong x = 0;

        public int Seed { get; private set; }

        public ilRand()
        {
            Seed = (int)(DateTime.Now.ToBinary() % uint.MaxValue);
            x = (ulong)Seed;
        }
        public ilRand(int seed)
        {
            Seed = seed;
            x = (ulong)Seed;
        }
        public ilRand(int seed, ulong x)
        {
            Seed = seed;
            x = x;
        }

        public void ChangeSeed()
        {
            Seed = (int)DateTime.Now.ToBinary();
            x = (ulong)Seed;
        }
        public void SetSeed(int seed)
        {
            Seed = seed;
            x = (ulong)Seed;
        }
        public void Initialize(int seed, ulong state)
        {
            Seed = seed;
            x = state;
        }

        public int Next()
        {
            return Next(0, int.MaxValue);
        }
        public int Next(int max)
        {
            return Next(0, max);
        }
        public int Next(int min, int max)
        {
            return (int)(next() % int.MaxValue) % (max - min) + min;
        }

        public double NextDouble()
        {
            return NextDouble(0, double.MaxValue);
        }
        public double NextDouble(double max)
        {
            return NextDouble(0, max);
        }
        public double NextDouble(double min, double max)
        {
            return (next() / (double)m) * (max - min) + min;
        }

        private ulong next()
        {
            x = ((x * a) % m);
            return x;
        }
    }
}

File: CPP_ilRand/ilRand.cpp
#include "ilRand.h"

#include <ctime>
#include <limits>


#define uint unsigned int

namespace Diploma2 {


	ilRand::ilRand()
	{
		seed = (int)(clock() % UINT_MAX);
		x = (ulonglong)(seed);
	}
	ilRand::ilRand(int s)
	{
		x = seed = s;
	}
	ilRand::ilRand(int s, int state)
	{
		seed = s;
		x = state;
	}


	ilRand::~ilRand()
	{
	}


	ulonglong ilRand::next() {
		x = ((x * a + c) % m);
		return x;
	}

	int ilRand::Next()
	{
		return Next(0, std::numeric_limits<int>::max());
	}
	int ilRand::Next(int max)
	{
		return Next(0, max);
	}
	int ilRand::Next(int min, int max)
	{
		return (int)(next() % std::numeric_limits<int>::max()) % (max - min) + min;
	}

	double ilRand::NextDouble()
	{
		return NextDouble(0, std::numeric_limits<double>::max());
	}
	double ilRand::NextDouble(double max)
	{
		return NextDouble(0, max);
	}
	double ilRand::NextDouble(double min, double max)
	{
		return (next() / (double)m) * (max - min) + min;
	}

	void ilRand::Initialize() {
		x = seed = clock();
	}
	void ilRand::Initialize(int s) {
		x = seed = s;
	}
	void ilRand::Initialize(int s, int state) {
		seed = s;
		x = state;
	}

}